### PR TITLE
Add `names` argument to ones, rand, randn, zeros, full; fix empty

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24087 Name inference for softmax, log_softmax and Dimname overloads.
* #24028 Fix out= function semantics for named tensors.
* #23973 Update tensor.view_names / tensor.names_ API
* #23962 Rename set_names -> view_names, set_names_ -> names_
* #23804 Implement tensor.align_to(names), torch.align_tensors(t1, t2, ...)
* #23808 Named inference for contiguous(), bernoulli variants, and dropout.
* #23746 Implement name inference rule for empty_like, clone
* **#23743 Add `names` argument to ones, rand, randn, zeros, full; fix empty**
* #23801 Improve test_namedtensor.py with named tensor equality check
* #23624 Add name propagation for at::alias, add tensor.set_names

In the short term, we implement this by having overloads for each of
these functions. In the long term, the plan is to move DimnameList to
TensorOptions so that we do not have to duplicate work.

Also fixes the implementation of empty. If there are no names, we should
just return an unnamed tensor instead of telling the user we don't
support their backend/layout.

Test Plan:
- [namedtensor ci]

gh-metadata: pytorch pytorch 23743 gh/zou3519/88/head

Differential Revision: [D16647820](https://our.internmc.facebook.com/intern/diff/D16647820)

### CODE REVIEW

**Review by nairbv — APPROVED** [review]

File: aten/src/ATen/native/TensorFactories.cpp [hunk 1]
```diff
@@ -122,6 +122,9 @@ Tensor empty(
     IntArrayRef size,
     at::optional<DimnameList> names,
     const TensorOptions& options) {
+  if (!names.has_value()) {
```
Comment (nairbv) [inline aten/src/ATen/native/TensorFactories.cpp]: I think you can just `if (! names)`

Comment (zou3519) [reply aten/src/ATen/native/TensorFactories.cpp]: They're the same in terms of performance (operator::bool is the same as has_value) and I wanted to be more explicit about it.
